build-tools/packages/build-cli/src/library/packageExports.ts

```diff
@@ -331,3 +331,57 @@ export function getTypesPathFromPackage(
 
 	return typesPath;
 }
+
+/**
+ * Read package "exports" to determine which of given "default" paths are present in map.
+ *
+ * @param packageJson - json content of package.json
+ * @param mapSrcQueryPathToOutKey - keys of map represent paths to match.
+ * @param emitDeclarationOnly - If true, "types" exports are considered.
+ * @param logger - optional Logger
+ * @returns Map with API tags or levels with export path data
+ */
+export function mapSourceExportsPath<TOutKey>(
+	packageJson: PackageJson,
+	mapSrcQueryPathToOutKey: ReadonlyMap<string | RegExp, TOutKey | undefined>,
+	emitDeclarationOnly: boolean,
+	logger?: Logger,
+): Map<TOutKey, ExportData> {
+	const mapKeyToOutput = new Map<TOutKey, ExportData>();
+
+	const { exports } = packageJson;
+
+	if (typeof exports !== "object" || exports === null) {
+		throw new Error('no valid "exports" within package properties');
+	}
+
+	// Iterate through exports looking for properties with values matching keys in map.
+	for (const [exportPath] of Object.entries(exports)) {
+		const resolvedExport = resolve.exports(packageJson, exportPath, {
+			conditions: emitDeclarationOnly ? [typesExportCondition] : [defaultExportCondition],
+		});
+		if (resolvedExport === undefined || resolvedExport.length === 0) {
+			throw new Error(`exports for ${exportPath} is undefined`);
+		}
+
+		const relPath = resolvedExport[0]
+			.replace(/(lib|dist)/g, "src")
+			.replace(/\.(d\.ts|js)$/, ".ts");
```

Comment:
Mostly already commented above, but this should not be done in here. The packageExports utilities should know/infer nothing about sources.